monai/data/dataset_calculator.py

```diff
@@ -0,0 +1,180 @@
+                the threshold to ``-1`` (default: ``0``).
+
+        """
+        voxel_sum = torch.as_tensor(0.0)
+        voxel_square_sum = torch.as_tensor(0.0)
+        voxel_max, voxel_min = [], []
+        voxel_ct = 0
+
+        for data in self.data_loader:
+            if self.image_key and self.label_key:
+                image, label = data[self.image_key], data[self.label_key]
+            else:
+                image, label = data
+
+            voxel_max.append(image.max().item())
+            voxel_min.append(image.min().item())
+
+            image_foreground = image[torch.where(label > foreground_threshold)]
+            voxel_ct += len(image_foreground)
+            voxel_sum += image_foreground.sum()
+            voxel_square_sum += torch.square(image_foreground).sum()
+
+        self.data_max, self.data_min = max(voxel_max), min(voxel_min)
+        self.data_mean = (voxel_sum / voxel_ct).item()
+        self.data_std = (torch.sqrt(voxel_square_sum / voxel_ct - self.data_mean ** 2)).item()
+
+    def calculate_percentiles(
+        self,
+        foreground_threshold: int = 0,
+        sampling_flag: bool = True,
+        interval: int = 10,
+        min_percentile: float = 0.5,
+        max_percentile: float = 99.5,
+    ):
+        """
+        This function is used to calculate the percentiles of intensities (and median) of the input dataset. To get
+        the required values, all voxels need to be accumulated. To reduce the memory used, this function can be set
+        to accumulate only a part of the voxels.
+
+        Args:
+            foreground_threshold: the threshold to distinguish if a voxel belongs to foreground, this parameter
+                is used to select the foreground of images for calculation. Normally, `label > 0` means the corresponding
+                voxel belongs to foreground, thus if you need to calculate the statistics for whole images, you can set
+                the threshold to ``-1`` (default: ``0``).
+            sampling_flag: whether to sample only a part of the voxels (default: ``True``).
+            interval: the sampling interval for accumulating voxels (default: ``10``).
+            min_percentile: minimal percentile (default: ``0.5``).
+            max_percentile: maximal percentile (default: ``99.5``).
+
+        """
+        all_intensities = []
+        for data in self.data_loader:
+            if self.image_key and self.label_key:
+                image, label = data[self.image_key], data[self.label_key]
+            else:
+                image, label = data
+
+            intensities = image[torch.where(label > foreground_threshold)].tolist()
+            if sampling_flag:
+                intensities = intensities[::interval]
+            all_intensities.append(intensities)
+
+        all_intensities = list(chain(*all_intensities))
+        self.data_min_percentile, self.data_max_percentile = np.percentile(
```

Comment:
We're still accumulating all the foreground pixel values from all the images here, unless we have some way of doing an incremental percentile calculation we are stuck with that.